File: tools/ultimate-csv/node/src/normalize.ts
import dayjs from 'dayjs';

export function toISODate(s: string, fmt: string): string {
  if (fmt === 'dd.MM.yyyy') {
    const m = s.match(/(\d{2})\.(\d{2})\.(\d{4})/);
    if (m) return `${m[3]}-${m[2]}-${m[1]}`;
  }
  if (/^\d{4}-\d{2}-\d{2}$/.test(s)) return s;
  const d = dayjs(s);
  return d.isValid() ? d.format('YYYY-MM-DD') : '';
}

export function toNumberDE(v: unknown): number {
  if (typeof v === 'number') return Number.isFinite(v) ? v : 0;
  if (typeof v !== 'string') return 0;
  let s = v.trim();
  if (!s) return 0;
  s = s.replace(/\s+/g, '');
  s = s.replace(/\u2212/g, '-');
  s = s.replace(/[€A-Za-z]/g, '');
  const neg = /^\(.*\)$/.test(s);
  if (neg) s = s.slice(1, -1);
  const lastComma = s.lastIndexOf(',');
  if (lastComma !== -1) {
    s = s.replace(/\./g, '');
    s = s.slice(0, lastComma).replace(/,/g, '') + '.' + s.slice(lastComma + 1);
  } else {
    const dots = (s.match(/\./g) || []).length;
    if (dots > 1) s = s.replace(/\./g, '');
  }
  s = s.replace(/^-?\./, (m) => (m.startsWith('-') ? '-0.' : '0.'));
  const n = Number(s);
  return Number.isFinite(n) ? (neg ? -n : n) : 0;
}

File: tools/ultimate-csv/python/ultimate_csv/normalize.py
def to_iso_date(s: str) -> str:
    s = s.strip()
    if len(s) == 10 and s[2] == '.' and s[5] == '.':
        return f"{s[6:10]}-{s[3:5]}-{s[0:2]}"
    return s

def to_number_de(v) -> float:
    if isinstance(v, (int, float)):
        return float(v)
    s = str(v).strip().replace('\u2212','-').replace('€','')
    s = s.replace(' ', '')
    if ',' in s:
        s = s.replace('.', '').replace(',', '.')
    try:
        return float(s)
    except:
        return 0.0


